compiler/rustc_ast/src/token.rs

```diff
@@ -884,7 +884,11 @@ pub enum NonterminalKind {
     PatWithOr,
     Expr,
     /// Matches an expression using the rules from edition 2021 and earlier.
-    Expr2021,
+    Expr2021 {
```

Comment:
unrelated -- but I wonder if we can come up with a more descriptive name. `expr_2021` gives me the vibe that its *new* to 2021 (i.e. >=2021) not legacy to 2021 (i.e. <=2021).

I quite like `PatParam` and `PatWithOr` naming scheme since it makes it less about edition and more about what it's parsing.

Reply:
Yeah, that would be awesome. I couldn't come up with a good succinct name yet, it's kinda hard lol ngl and the grammar may still change, too (cc #126697). Note that we're already tracking this under *Unresolved Questions* in #123742.

Reply:
I agree the `_2021` suffix's meaning is non-obvious, and even after learning what it means it I keep having trouble remembering the meaning.

Reply:
Maybe we can use the year as a suffix for saying "the moment in was introduced", that would mean we would have `Expr2015` and `Expr2024`; that would follow "the vibe that its new to 2024".

Reply:
Just to confirm, are we agreeing that Expr2021 would include all features up to the 2021 edition, and Expr would continue to represent the current edition's features?

Or are you proposing a slightly different idea from what we are currently implementing?

By the way, I am happy with either approach, but I think it is important for users that we clearly indicate the edition where the expr_* is necessary to maintain old behavior.